### PR TITLE
Equations: do all unification at start-up

### LINKED ISSUE #3188: Move all equation-related unification to start-up

### PR DESCRIPTION
**Pull request by ana-pantilie**
Fixes #3188 

Tested performance of https://github.com/runtimeverification/haskell-backend/issues/3182#issuecomment-1205482619, observed a 10x improvement!

Ran a part of the KEVM test suite locally, there don't seem to be any functional regressions. Didn't compare performance.

### Scope:
- simplify the `argument` and `antiLeft` of equations at initialization, removing the cost of doing that during every attempt to apply equations
- after processing the equations, they are all applied using the matching procedure
- adds support for testing this pre-processing of equations, removes tests which tested applying equations using unification

### Estimate:

---

###### Review checklist

The author performs the actions on the checklist. The reviewer evaluates the work and checks the boxes as they are completed.

-   [ ] **Summary.** Write a summary of the changes. Explain what you did to fix the issue, and why you did it. Present the changes in a logical order. Instead of writing a summary in the pull request, you may push a clean Git history.
-   [ ] **Documentation.** Write documentation for new functions. Update documentation for functions that changed, or complete documentation where it is missing.
-   [ ] **Tests.** Write unit tests for every change. Write the unit tests that were missing before the changes. Include any examples from the reported issue as integration tests.
-   [ ] **Clean up.** The changes are already clean. Clean up anything near the changes that you noticed while working. This does not mean only spatially near the changes, but logically near: any code that interacts with the changes!

### CODE REVIEW

File: kore/src/Kore/Attribute/Axiom.hs [hunk 1]
```diff
@@ -295,6 +295,7 @@ getPriorityOfAxiom
                 defaultSimplificationPriority
             (Priority Nothing, Owise False, IsSimplification (Just value)) ->
                 value
+            -- TODO: remove
```
Comment (ana-pantilie) [inline kore/src/Kore/Attribute/Axiom.hs:298]:
```suggestion
```

**Comment by ehildenb** [comment]
KEVM performance:

```
* 43e3e7113 - !!! (22232ca31b) deps/k: 22232ca31b - !!! (origin/fix-equation-unification) haskell-backend/src/main/native/haskell-
| 0  4066.12s  24983.45s  680.35s  6872348KB  make  test-prove        TEST_SYMBOLIC_BACKEND=haskell  KEVM='profile  .build/logs/prove-haskell     timeout  5000  kevm'  -j6  -k  KPROVE_OPTS='--bug-report'

* 82576bc98 - Storage tests new (#1319)                                                        - Raoul (upstream/master)
| 0  4157.74s  22508.85s  507.57s  3965760KB  make  test-prove        TEST_SYMBOLIC_BACKEND=haskell  KEVM='profile  .build/logs/prove-haskell     timeout  5000  kevm'  -j6  -k  KPROVE_OPTS='--bug-report'
| post: commit=$(git rev-parse --short=8 HEAD) && mkdir -p logs/${commit} && rm -rf logs/${commit}/profile && mv .build/logs logs/${commit}/profile && mv kevm-bug-*.tar.gz logs/${commit}/profile
```

Before 4157s, after 4066s (overall time, 6 tests in parallel).

Individual test breakdown:

```
 Test                                            | 43e3e711-prove-haskell time | b9580ca4-prove-haskell time | (43e3e711-prove-haskell/b9580ca4-prove-haskell) time 
-------------------------------------------------+-----------------------------+-----------------------------+------------------------------------------------------
 specs/functional/storageRoot-spec.k             | 32.33                       | 36.08                       | 0.8960643015521065                                   
 specs/examples/solidity-code-spec.md            | 178.07                      | 193.89                      | 0.9184073443705194                                   
 specs/mcd/flipper-addu48u48-fail-rough-spec.k   | 96.5                        | 104.95                      | 0.9194854692710814                                   
 specs/bihu/forwardToHotWallet-failure-3-spec.k  | 320.88                      | 348.65                      | 0.9203499211243368                                   
 specs/benchmarks/keccak00-spec.k                | 231.87                      | 251.54                      | 0.9218017015186452                                   
 erc20/ds/balanceOf-spec.k                       | 189.38                      | 205.36                      | 0.922185430463576                                    
 specs/bihu/forwardToHotWallet-failure-4-spec.k  | 551.87                      | 597.64                      | 0.9234154340405596                                   
 specs/benchmarks/requires01-a0gt0-spec.k        | 136.86                      | 148.12                      | 0.9239805563056982                                   
 specs/benchmarks/encodepacked-keccak01-spec.k   | 190.46                      | 206.01                      | 0.9245182272705209                                   
 specs/bihu/forwardToHotWallet-failure-2-spec.k  | 291.86                      | 315.46                      | 0.9251886134533698                                   
 erc20/ds/approve-failure-spec.k                 | 114.35                      | 123.56                      | 0.9254613143412107                                   
 specs/benchmarks/storagevar02-nooverflow-spec.k | 121.49                      | 131.23                      | 0.925779166349158                                    
 erc20/hkg/totalSupply-spec.k                    | 44.19                       | 47.72                       | 0.9260268231349539                                   
 erc20/ds/transfer-failure-2-a-spec.k            | 539.21                      | 581.48                      | 0.9273061842195777                                   
 specs/mcd/vat-mului-pass-spec.k                 | 273.67                      | 294.99                      | 0.927726363605546                                    
 specs/mcd/dai-symbol-pass-spec.k                | 227.92                      | 245.64                      | 0.9278619117407588                                   
 specs/mcd/cat-file-addr-pass-rough-spec.k       | 247.52                      | 266.56                      | 0.9285714285714286                                   
 specs/benchmarks/overflow00-overflow-spec.k     | 134.51                      | 144.84                      | 0.9286799226732946                                   
 specs/benchmarks/encode-keccak00-spec.k         | 373.9                       | 402.34                      | 0.929313515931799                                    
 specs/mcd/end-subuu-pass-spec.k                 | 138.72                      | 149.17                      | 0.9299456995374406                                   
 specs/benchmarks/overflow00-nooverflow-spec.k   | 140.4                       | 150.82                      | 0.9309110197586528                                   
 erc20/hkg/balanceOf-spec.k                      | 99.87                       | 107.21                      | 0.9315362372912975                                   
 specs/mcd/flopper-cage-pass-spec.k              | 245.05                      | 262.95                      | 0.931926221715155                                    
 specs/benchmarks/staticarray00-spec.k           | 190.31                      | 204.15                      | 0.9322067107518981                                   
 specs/mcd/flipper-tau-pass-spec.k               | 161.19                      | 172.89                      | 0.9323269130661115                                   
 specs/benchmarks/ecrecover00-siginvalid-spec.k  | 296.22                      | 317.52                      | 0.9329176114890402                                   
 specs/mcd/vat-subui-pass-rough-spec.k           | 371.03                      | 397.7                       | 0.932939401558964                                    
 specs/benchmarks/staticloop00-a0lt10-spec.k     | 244.07                      | 261.41                      | 0.9336674189969778                                   
 erc20/ds/totalSupply-spec.k                     | 115.12                      | 123.28                      | 0.9338092147955873                                   
 specs/functional/lemmas-spec.k                  | 24.75                       | 26.5                        | 0.9339622641509434                                   
 specs/mcd/dai-adduu-fail-rough-spec.k           | 78.13                       | 83.61                       | 0.9344576007654586                                   
 specs/mcd/flipper-ttl-pass-spec.k               | 159.06                      | 170.14                      | 0.934877159985894                                    
 specs/mcd/dsvalue-peek-pass-rough-spec.k        | 308.08                      | 329.28                      | 0.935617103984451                                    
 specs/benchmarks/requires01-a0le0-spec.k        | 102.02                      | 109.03                      | 0.9357057690543886                                   
 specs/mcd/vat-subui-fail-rough-spec.k           | 268.94                      | 287.35                      | 0.9359317904993909                                   
 erc20/ds/transferFrom-failure-1-d-spec.k        | 448.11                      | 478.67                      | 0.9361564334510205                                   
 erc20/ds/transferFrom-failure-2-c-spec.k        | 295.68                      | 315.5                       | 0.9371790808240887                                   
 specs/mcd/dsvalue-read-pass-summarize-spec.k    | 146.38                      | 156.17                      | 0.9373119036946918                                   
 specs/mcd/cat-exhaustiveness-spec.k             | 156.83                      | 167.28                      | 0.9375298900047825                                   
 erc20/ds/allowance-spec.k                       | 285.95                      | 304.91                      | 0.9378177167032894                                   
 erc20/ds/transfer-failure-1-c-spec.k            | 380.33                      | 405.53                      | 0.9378590979705571                                   
 specs/benchmarks/storagevar00-spec.k            | 100.72                      | 107.39                      | 0.9378899338858366                                   
 erc20/hkg/allowance-spec.k                      | 135.05                      | 143.99                      | 0.9379123550246545                                   
 erc20/ds/approve-success-spec.k                 | 509.43                      | 542.88                      | 0.9383841732979664                                   
 erc20/hkg/transfer-failure-2-spec.k             | 186.17                      | 198.38                      | 0.9384514568000806                                   
 specs/mcd/vat-deny-diff-fail-rough-spec.k       | 451.55                      | 480.73                      | 0.9393006469327897                                   
 specs/benchmarks/ecrecover00-sigvalid-spec.k    | 378.72                      | 403.11                      | 0.9394954230855102                                   
 specs/mcd/vat-addui-fail-rough-spec.k           | 269.08                      | 286.12                      | 0.9404445687124283                                   
 erc20/ds/transfer-failure-2-b-spec.k            | 271.19                      | 288.11                      | 0.9412724306688417                                   
 specs/mcd/vat-subui-pass-spec.k                 | 378.36                      | 401.85                      | 0.941545352743561                                    
 specs/benchmarks/functional-spec.k              | 42.3                        | 44.87                       | 0.9427234232226431                                   
 specs/mcd/dsvalue-read-pass-spec.k              | 332.15                      | 352.09                      | 0.9433667528188815                                   
 erc20/hkg/approve-spec.k                        | 169.54                      | 179.5                       | 0.9445125348189415                                   
 specs/benchmarks/storagevar01-spec.k            | 112.2                       | 118.79                      | 0.9445239498274265                                   
 specs/mcd/vow-flog-fail-rough-spec.k            | 398.12                      | 421.45                      | 0.9446434927037609                                   
 specs/bihu/forwardToHotWallet-failure-1-spec.k  | 214.41                      | 226.77                      | 0.94549543590422                                     
 specs/benchmarks/storagevar02-overflow-spec.k   | 118.25                      | 124.96                      | 0.9463028169014085                                   
 specs/benchmarks/bytes00-spec.k                 | 220.02                      | 232.25                      | 0.9473412271259419                                   
 specs/mcd/vat-move-diff-rough-spec.k            | 806.67                      | 850.53                      | 0.9484321540686396                                   
 erc20/hkg/transferFrom-failure-2-spec.k         | 419.37                      | 441.98                      | 0.9488438390877415                                   
 specs/benchmarks/storagevar03-spec.k            | 87.19                       | 91.82                       | 0.949575255935526                                    
 specs/mcd/vow-fess-fail-rough-spec.k            | 430.71                      | 450.99                      | 0.9550322623561497                                   
 specs/mcd/dstoken-burn-self-fail-rough-spec.k   | 2156.55                     | 2257.51                     | 0.9552781604511165                                   
 specs/examples/sum-to-n-spec.k                  | 27.78                       | 29.07                       | 0.95562435500516                                     
 erc20/hkg/transferFrom-success-2-spec.k         | 541.65                      | 566.6                       | 0.9559654076950229                                   
 specs/functional/merkle-spec.k                  | 20.11                       | 21.01                       | 0.9571632555925749                                   
 specs/benchmarks/address00-spec.k               | 130.45                      | 136.07                      | 0.9586977291100168                                   
 erc20/hkg/transfer-success-2-spec.k             | 233.31                      | 243.09                      | 0.9597679871652475                                   
 erc20/hkg/transfer-failure-1-spec.k             | 314.74                      | 327.48                      | 0.9610968608770001                                   
 erc20/hkg/transfer-success-1-spec.k             | 387.16                      | 402.34                      | 0.962270716309589                                    
 specs/mcd/functional-spec.k                     | 25.59                       | 26.55                       | 0.9638418079096045                                   
 specs/functional/infinite-gas-spec.k            | 21.67                       | 22.47                       | 0.9643969737427682                                   
 TOTAL                                           | 18843.290000000005          | 20047.879999999997          | 0.9399143450579317    
```

Speedup on every test, ranging from 3.5% to 10.4%.

File: kore/test/Test/Kore/Equation/Application.hs [hunk 2]
```diff
@@ -355,200 +354,6 @@ test_attemptEquation =
         (Pattern.fromTermLike a)
     ]
 
-test_attemptEquationUnification :: [TestTree]
```
Comment (goodlyrottenapple) [inline kore/test/Test/Kore/Equation/Application.hs:358]: Why are all these tests being removed? Do they no longer apply?

Comment (goodlyrottenapple) [reply kore/test/Test/Kore/Equation/Application.hs:358]: oh i guess these are to do with 
> removes tests which tested applying equations using unification

Comment (ana-pantilie) [reply kore/test/Test/Kore/Equation/Application.hs:358]: They used to duplicate the other tests in that module which tested applying equations by matching.

File: kore/test/Test/Kore/Rewrite/Function/Integration.hs [hunk 3]
```diff
@@ -1111,21 +677,8 @@ withApplied check comment rules term =
...
-        check actual
-
 applies
-    , notApplies
-    , appliesUnification
-    , notAppliesUnification ::
+    , notApplies ::
         TestName ->
         [Equation RewritingVariableName] ->
         TermLike RewritingVariableName ->
```
Comment (jberthold) [inline kore/test/Test/Kore/Rewrite/Function/Integration.hs:684]: (nit) maybe in one line, now that there are only two?

File: kore/test/Test/Kore/Simplify/Integration.hs [hunk 4]
```diff
@@ -741,31 +726,29 @@ test_simplificationIntegrationUnification =
...
-                    )
+                                (mkElemVar Mock.yConfig)
+                            ]
+                        )
+                    ]
                 )
                 initial
         assertEqual "" expect actual
-    , testCase "function application with top predicate" $ do
+    , testCase "TESTING function application with top predicate" $ do
```
Comment (jberthold) [inline kore/test/Test/Kore/Simplify/Integration.hs:751]: Debugging left-over?
```suggestion
    , testCase "function application with top predicate" $ do
```

**Review by jberthold — APPROVED** [review]
LGTM, maybe try my refactoring suggestion in a follow-up so we get the faster version on master quickly.

**Review by goodlyrottenapple — APPROVED** [review]
LGTM

File: kore/test/Test/Kore/Simplify/Integration.hs (hunk 4 above)
Comment (ana-pantilie) [reply kore/test/Test/Kore/Simplify/Integration.hs:751]: Yep, thanks for spotting this!

**Pull request merged**